Sources/NIOSSL/SSLCertificateExtensions.swift

```diff
@@ -0,0 +1,197 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+@_implementationOnly import CNIOBoringSSL
+@_implementationOnly import CNIOBoringSSLShims
+
+public struct _SSLCertificateExtensions {
+    private enum Storage {
+        final class Deallocator {
+            let reference: OpaquePointer!
+            
+            init(takeOwnershipOf reference: OpaquePointer!) {
+                self.reference = reference
+            }
+            
+            deinit {
+                CNIOBoringSSL_sk_X509_EXTENSION_free(self.reference)
+            }
+        }
+        
+        case owned(Deallocator)
+        case borrowed(reference: OpaquePointer!, owner: AnyObject)
```

Comment:
It either needs to be optional/IUO or we need to handle the empty case differently [because `CNIOBoringSSL_X509_get0_extensions` returns `nil` if no extensions are present](https://www.openssl.org/docs/man1.1.1/man3/X509_get0_extensions.html).
The good thing is that `CNIOBoringSSL_sk_X509_EXTENSION_num` returns `0` if we give it a `nil` pointer so we should be safe. That being said, I'm wondering if we should rely on that behaviour or if we should add our own additional guard. WDYT?

Reply:
In this case we don't want it to be an IUO, right, we want it to be an actual optional.

Reply:
fixed in https://github.com/apple/swift-nio-ssl/pull/359/commits/de44d779536fcbb48e0b4b698da05e9439f2b5a6